jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java

```diff
@@ -65,47 +67,57 @@ public class Jcmd {
 	 * @param args Application arguments.  See help text for details.
 	 */
 	public static void main(String[] args) {
-		String command = null;
-		/* An empty argument list is a request for a list of VMs */
+		if ((args.length == 1) && Arrays.stream(HELP_OPTIONS).anyMatch(args[0]::equals)) {
+			System.out.printf(HELPTEXT);
+			return;
+		}
+
+		List<AttachProvider> providers = AttachProvider.providers();
+		if ((providers == null) || providers.isEmpty()) {
+			System.err.println("no attach providers available"); //$NON-NLS-1$
+			return;
+		}
+
+		AttachProvider openj9Provider = providers.get(0);
+		List<VirtualMachineDescriptor> vmds = openj9Provider.listVirtualMachines();
+		if ((vmds == null) || vmds.isEmpty()) {
+			System.err.println("no vm found"); //$NON-NLS-1$
+			return;
+		}
+
+		/* An empty argument list is a request for a list of VMs. */
 		final String firstArgument = (0 == args.length) ? "-l" : args[0]; //$NON-NLS-1$
 		if ("-l".equals(firstArgument)) { //$NON-NLS-1$
-			List<AttachProvider> providers = AttachProvider.providers();
-			AttachProvider myProvider = null;
-			if (!providers.isEmpty()) {
-				myProvider = providers.get(0);
-			}
-			if (null == myProvider) {
-				System.err.println("no attach providers available"); //$NON-NLS-1$
-			} else {
-				for (VirtualMachineDescriptor vmd : myProvider.listVirtualMachines()) {
-					StringBuilder outputBuffer = new StringBuilder(vmd.id());
-					Util.getTargetInformation(myProvider, vmd, false, false, true, outputBuffer);
-					System.out.println(outputBuffer.toString());
-				}
+			for (VirtualMachineDescriptor vmd : vmds) {
+				StringBuilder outputBuffer = new StringBuilder(vmd.id());
+				Util.getTargetInformation(openj9Provider, vmd, false, false, true, outputBuffer);
+				System.out.println(outputBuffer.toString());
 			}
-		} else if ((args.length == 1) && (null != firstArgument) && Arrays.stream(HELP_OPTIONS).anyMatch(firstArgument::equals)) {	
-			System.out.printf(HELPTEXT);
 		} else {
-			command = DiagnosticUtils.makeJcmdCommand(args, 1);
+			String command = DiagnosticUtils.makeJcmdCommand(args, 1);
 			if (command.isEmpty()) {
 				System.err.printf("There is no jcmd command.%n"); //$NON-NLS-1$
 				System.out.printf(HELPTEXT);
 			} else {
-				String vmid = firstArgument;
 				AttacherDiagnosticsProvider diagProvider = new AttacherDiagnosticsProvider();
-				try {
-					diagProvider.attach(vmid);
+				List<String> vmids = Util.findMatchVMIDs(vmds, firstArgument);
+				for (String vmid : vmids) {
+					System.out.println(vmid + ":");
```

Comment:
This could be skipped if the vmid is equal to the first argument, to keep the output compatible with the old behavior.
$NON-NLS-1$ is missing.

Reply:
Added a check for `firstArgument` and `$NON-NLS-1$`.